README.md
# 2018 November
[![Build Status](https://travis-ci.org/ebabel-games/2018-november.svg?branch=master)](https://travis-ci.org/ebabel-games/2018-november) [![dependencies Status](https://david-dm.org/ebabel-games/2018-november.svg)](https://david-dm.org/ebabel-games/2018-november.svg)

Collect all the stars as fast as possible! More and more deadly bombs will appear.

![Screenshot of gameplay](assets/screenshots/2018-11-12-2325.png)

### Requirements
We need [Node.js](https://nodejs.org) to install and run scripts.

## Install and run
Run next commands in your terminal:

| Command | Description |
|---------|-------------|
| `npm install` | Install dependencies.|
| `npm test` | Lint the source code and run all automated tests.|
| `npm run build` | One-off build of the game.|
| `npm run watch` | Continuously build the game during development.|
| `npm start` | Build bundle and run game on localhost:8080. <br> Press `Ctrl + c` to kill **http-server** process. |

Browse to http://localhost:8080 to play the game.

## Deploy to Firebase
```
npm install -g firebase-tools
firebase login
firebase use --add [name of your Firebase app]
firebase deploy --only hosting
```

## Test localhost on a touch device
- [Check your local IP](https://www.whatismyip.com/) on your network (it should start with 192.168.).
- Put your touch device on the same network as your development machine (same Wifi for eample).
- Add the port :8080 to the IP.
- You should now be able to test the game running on your machine with your touch device.

## Switch the debug mode on or off
By default, debug mode is off. To switch it on, run in Javascript console of your browser:
```
localStorage['debug'] = true;
```

## Acceptance criterias
For each mini-game made in just 1 month, I want to hit as many as possible of these targets:
- [x] Player interactions are central to the game.
- [x] There is a clear outcome to the game: the player either wins or loses.
- [x] Graphics and animations are unique and created for the game.
- [ ] Sounds are unique and created for the game. (is not, at least Open source and Royalty Free)
- [x] Layout is responsive.
- [x] Single page app: It's possible to play the game multiple times without reloading the page.
- [x] Portrait and landscape modes are both supported.
- [ ] Playable on the most popular (NL) touch device browsers at that time:
  - [ ] Edge 17,
  - [ ] Firefox 62,
  - [ ] Chrome 69,
  - [ ] Chrome 70,
  - [ ] and Safari 12.
- [ ] Playable on the most popular (NL) desktop browsers at that time:
  - [ ] ~~IE 11~~ (not supported despite popularity),
  - [x] Edge 17,
  - [x] Firefox 62,
  - [x] Chrome 69,
  - [x] Chrome 70,
  - [x] and Safari 12.
- [x] Performance is stable enough with a good frame rate throughout the gaming experience (60 fps).
- [x] Unit tests, test coverage reporting, and linting are all setup.
- [ ] Test coverage is higher than 80%.
- [x] Travis CI build pipeline passes.
- [x] Semantic versioning and release tags standards are adhered to.
- [x] Published on [Firebase](https://firebase.google.com) to support HTTPS.
- [x] Listed on my [eBabel gaming website](https://ebabel.eu).
- [x] Works offline as a Progressive Web App (PWA) with CacheStorage.
- [x] Can be updated if a new version is made.
- [ ] Can be installed as a [Desktop Progressive Web App](https://developers.google.com/web/progressive-web-apps/desktop) with:
  - [ ] ~~IE 11~~ (not supported despite popularity),
  - [ ] Edge 17,
  - [ ] Firefox 62,
  - [x] Chrome 69 on [x] Windows 10, [ ] MacOS 10.4, [ ] iOS 12, and [ ] Android 9.0,
  - [x] Chrome 70 on [x] Windows 10, [ ] MacOS 10.4, [ ] iOS 12, and [ ] Android 9.0,
  - [ ] and Safari 12.
- [ ] Can be [added to homescreen](https://developers.google.com/web/fundamentals/app-install-banners/) (A2HS) on smartphone with:
  - [ ] Edge 17,
  - [ ] Firefox 62,
  - [ ] Chrome 69,
  - [ ] Chrome 70,
  - [ ] and Safari 12.
- [x] Published on [itch.io](https://ebabel.itch.io/nov-2018)
- [x] Published on [One Game a Month](http://www.onegameamonth.com/eBabel)
- [x] Published on [New Grounds](https://thomasamar.newgrounds.com/)



CHANGELOG.md
# 2018 November game

## 1.7.0
- Publish game on [itch.io](https://ebabel.itch.io/nov-2018)

## 1.6.0
- Setup Desktop Progressive Web App for the game.
- See https://developers.google.com/web/progressive-web-apps/desktop

## 1.5.0
- Host game on Firebase instead of Github Pages.

## 1.4.0
- Add Progressive Web App (PWA) features: offline, cache, updates of new version, installable on mobile homescreens, and on desktops.
- See O'Reilly book [Building Progressive Web Apps](https://www.oreilly.com/library/view/building-progressive-web/9781491961643/) by [Tal Ater](https://twitter.com/talater) for more info.

## 1.3.0
- In debug mode, stars and platforms can be manually dragged to new positions.

## 1.2.1
- Fix compatibility issues for supported browsers.

## 1.2.0
- Explore beyond the initial visual game screen, like Super Mario who sees more than first screen.
- See [Zenva course](https://academy.zenva.com/lesson/arcade-basics/) for more info.

## 1.1.0
- Touch enabled for mobile devices.
- Code based on Phaser 3 release documentation for [handling touch events](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/touchevents/).

## 1.0.0
- First version of the game, launch for desktop browsers with keyboard input (arrow keys left, right, and up).
- Project started from a [boilerplate](https://github.com/photonstorm/phaser3-project-template).
- See [tutorial walkthrough](http://phaser.io/tutorials/making-your-first-phaser-3-game) and [Phaser 3 book](https://phaser.io/news/2018/07/brand-new-phaser-3-book) for more info.


src/constants.js
'use strict';

// Color palette. For more info, see /COLOR_PALETTE.md documentation.
const colors = [
  ['#58BBDC', '#B6E6F7', '#81D0EA', '#36A6CB', '#168DB5'],  // Primary color.
  ['#FFD060', '#FFEBB9', '#FFDC88', '#FFC63D', '#FFBB16'],  // Secondary color (1).
  ['#FF7960', '#FFC4B9', '#FF9A88', '#FF5B3D', '#FF3A16'],  // Secondary color (2).
];

const colors0x = [
  [0x58BBDC, 0xB6E6F7, 0x81D0EA, 0x36A6CB, 0x168DB5],  // Primary color.
  [0xFFD060, 0xFFEBB9, 0xFFDC88, 0xFFC63D, 0xFFBB16],  // Secondary color (1).
  [0xFF7960, 0xFFC4B9, 0xFF9A88, 0xFF5B3D, 0xFF3A16],  // Secondary color (2).
];

const debug = (localStorage['debug'] === 'true') ? true : false;

module.exports = {
  debug,

  scoreDefault: 0,
  scoreCollectStar: 10,

  // The game saves performance by not running code until it stops getting called for at least x milliseconds.
  debounceDelay: 250,

  // The game needs the player to notice something and will not run for x milliseconds.
  minimumDelay: 1500,

  colors,
  colors0x,

  headerFontFamily: '\'Gill Sans\', \'Gill Sans MT\', Calibri, \'Trebuchet MS\', sans-serif',

  gameWidth: 800,
  gameHeight: 600,
  gravity: 300,

  // World bounds can be greater than canvas gameWidth and gameHeight,
  // so the camera can move to new areas beyond the initial screen area.
  worldBoundsWidth: 800,
  worldBoundsHeight: 1200,

  skyKey: 'sky',
  skyAsset: 'assets/sky.svg',
  skyPositionX: 400,
  skyPositionY: 300,

  starsKey: 'star',
  starsAsset: 'assets/star.svg',
  starsRepeat: 11,
  starsPositionX: 14,
  starsPositionY: 60,
  starsPositionStepX: 70,
  starsMinBounceY: 0.2,
  starsMaxBounceY: 0.4,
  starsThresholdSpawnBomb: 0,
  stars: [
    // Ground.
    [154, 1157],
    [294, 1157],

    // Lowest block.
    [504, 1008],
    [574, 1008],
    [644, 1008],
    [14,  913],
    [84,  913],
    [364, 855],
    [434, 855],
    [714, 825],
    [784, 825],
    [224, 813],

    // Middle block.
    [504, 708],
    [574, 708],
    [644, 708],
    [14,  613],
    [84,  613],
    [364, 555],
    [434, 555],
    [714, 525],
    [784, 525],
    [224, 513],

    // Upper block.
    [504, 408],
    [574, 408],
    [644, 408],
    [14,  313],
    [84,  313],
    [364, 255],
    [434, 255],
    [714, 225],
    [784, 225],
    [224, 213],
  ],

  platformKey: 'platform',
  platformAsset: 'assets/platform.svg',
  platformFrameWidth: 32,
  platformFrameHeight: 32,
  platformIsStatic: true,
  // platformsX, platformsY, platformsWidth, platformsHeight, and platformsFrame.
  platformsX: 0,
  platformsY: 1,
  platformsWidth: 2,
  platformsHeight: 3,
  platformsFrame: 4,
  platforms: [
    // Ground.
    [400, 1184, 32 * 25, 32, 1],

    // Lowest block.
    [462, 1035, 32, 32, 0],   // 462 = 590 - (32 * 7 / 2) - (32 / 2)
    [590, 1035, 224, 32, 1],  // 224 = 32 * 7
    [718, 1035, 32, 32, 2],   // 718 = 590 + (32 * 7 / 2) + (32 / 2)

    [16, 940, 32, 32, 0],
    [48, 940, 32, 32, 1],
    [80, 940, 32, 32, 2],
    
    [336, 882, 32, 32, 0],
    [400, 882, 96, 32, 1],
    [464, 882, 32, 32, 2],

    [688, 852, 32, 32, 0],
    [736, 852, 64, 32, 1],
    [784, 852, 32, 32, 2],

    [188, 840, 32, 32, 0],
    [220, 840, 32, 32, 1],
    [252, 840, 32, 32, 2],

    // Middle block.
    [462, 735, 32, 32, 0],
    [590, 735, 224, 32, 1],
    [718, 735, 32, 32, 2],

    [16, 640, 32, 32, 0],
    [48, 640, 32, 32, 1],
    [80, 640, 32, 32, 2],
    
    [336, 582, 32, 32, 0],
    [400, 582, 96, 32, 1],
    [464, 582, 32, 32, 2],

    [688, 552, 32, 32, 0],
    [736, 552, 64, 32, 1],
    [784, 552, 32, 32, 2],

    [188, 540, 32, 32, 0],
    [220, 540, 32, 32, 1],
    [252, 540, 32, 32, 2],

    // Upper block.
    [462, 435, 32, 32, 0],
    [590, 435, 224, 32, 1],
    [718, 435, 32, 32, 2],

    [16, 340, 32, 32, 0],
    [48, 340, 32, 32, 1],
    [80, 340, 32, 32, 2],
    
    [336, 282, 32, 32, 0],
    [400, 282, 96, 32, 1],
    [464, 282, 32, 32, 2],

    [688, 252, 32, 32, 0],
    [736, 252, 64, 32, 1],
    [784, 252, 32, 32, 2],

    [188, 240, 32, 32, 0],
    [220, 240, 32, 32, 1],
    [252, 240, 32, 32, 2],
  ],
  platformsFrameKeyLeft: 0,
  platformsFrameKeyMiddle: 1,
  platformsFrameKeyRight: 2,

  playerKey: 'hero',
  playerAsset: 'assets/hero.png',
  playerFrameWidth: 32,
  playerFrameHeight: 48,
  playerStartX: 120,
  playerStartY: 50,
  playerGravity: 30,
  playerBounce: 0.2,
  playerCollideWorldBounds: true,
  playerLeftVelocityX: -160,
  playerRightVelocityX: 160,
  playerTurnVelocityX: 0,
  playerUpVelocityY: -330,
  playerDeadTint: colors0x[1][0],
  playerMaxY: 1144,

  playerAnimations: {
    left: {
      key: 'left',
      frames: [0, 1, 2, 3],
      frameRate: 10,
      repeat: -1,
    },
    turn: {
      key: 'turn',
      frame: 4,
      frameRate: 20,
    },
    right: {
      key: 'right',
      frames: [0, 1, 2, 3],
      frameRate: 10,
      repeat: -1,
    },
  },

  scoreTextPositionX: 8,
  scoreTextPositionY: 4,
  scoreTextDefault: 'Score: 0',
  scoreTextFontSize: '32px',
  scoreTextFill: colors[1][0],

  gameOverTextPositionX: 132,
  gameOverTextPositionY: 200,
  gameOverTextDefault: 'Game Over',
  gameOverTextFontSize: '120px',
  gameOverTextFill: colors[1][3],
  gameOverTextStroke: colors[2][0],
  gameOverTextStrokeThickness: 12,

  winTextPositionX: 155,
  winTextPositionY: 200,
  winTextDefault: 'You won!',
  winTextFontSize: '120px',
  winTextFill: colors[1][3],
  winTextStroke: colors[2][0],
  winTextStrokeThickness: 12,

  bombLeftEdge: 0,
  bombMiddleBorder: 400,
  bombRightEdge: 800,
  bombKey: 'bomb',
  bombAsset: 'assets/bomb.svg',
  bombStartPositionY: 16,
  bombBounce: 1,
  bombCollideWorldBounds: true,
  bombVelocityMinLower: -200,
  bombVelocityMaxLower: 200,
  bombVelocityMax: 20,
  bombAllowGravity: false,
  bombIsCircle: true,
  bombSpawnInterval: (debug) ? 150000 : 15000,

  audioCollectStarKey: 'collect-star',
  audioCollectStarAssets: ['assets/kenney-sounds/coin1.ogg', 'assets/kenney-sounds/coin1.mp3'],
  audioExplosionKey: 'explosion',
  audioExplosionAssets: ['assets/kenney-sounds/explosion1.ogg', 'assets/kenney-sounds/explosion1.mp3'],
  audioGameOverKey: 'gameover',
  audioGameOverAssets: ['assets/kenney-sounds/gameover3.ogg', 'assets/kenney-sounds/gameover3.mp3'],
  audioWinKey: 'win',
  audioWinAssets: ['assets/kenney-sounds/jingles_NES03.ogg', 'assets/kenney-sounds/jingles_NES03.mp3'],
};
